tests/helpers/rabbitMQ.ts

```diff
@@ -0,0 +1,34 @@
+/* tslint:disable:no-console */
+/* tslint:disable:no-relative-imports */
+import { connect, ConsumeMessage, Connection } from 'amqplib'
+
+export const RabbitMQ = async (rabbitUrl: string) => {
+  console.log(`Connecting to RabbitMQ...`)
+  const connection = await connect(rabbitUrl)
+
+  return {
+    consume: (
```

Comment:
We have been declaring the functions used by factories inside the factories themselves:

```ts
export const RabbitMQ = async (rabbitUrl: string) => {
  console.log(`Connecting to RabbitMQ...`)
  
  const connection = await connect(rabbitUrl)

  const consume = async ( // doesn't take connection anymore, grabs it from lexical context
    exchange: string,
    consume: (message: ConsumeMessage) => Promise<void>,
  ) => {
    const channel = await connection.createChannel()
    await channel.assertExchange(exchange, 'fanout', { durable: false })
    const assertQueue = await channel.assertQueue('', { exclusive: true })
    channel.bindQueue(assertQueue.queue, exchange, '')
    channel.consume(assertQueue.queue, consume)
  }

  const stop = async () => {
    console.log(`Closing RabbitMQ connection...`)
    await connection.close()
  }

  return {
    consume,
    stop,
  }
}
```

This keeps the `return` nice and tidy, making it immediately clear what is being made public.

Also: do declare an interface for factories:

```ts
type ConsumeCallback = (message: ConsumeMessage) => Promise<void>

interface RabbitMQ {
  readonly consume: (exchange: string, callback: ConsumeCallback) => Promise<void>
  readonly stop: () => Promise<void>
}

export const RabbitMQ = async (rabbitUrl: string): RabbitMQ => {
```

Reply:
If you feel strongly about not placing the declaration bodies of `consume` and `stop` inside the factory, try currying them:

```ts
export const RabbitMQ = async (rabbitUrl: string) => {
  console.log(`Connecting to RabbitMQ...`)
  const connection = await connect(rabbitUrl)
  return {
    consume: consume(connection),
    close: close(connection),
  }
}

const consume = (connection: Connection) => async (
    exchange: string,
    consume: (message: ConsumeMessage) => Promise<void>,
  ) => {
  const channel = await connection.createChannel()
  await channel.assertExchange(exchange, 'fanout', { durable: false })
  const assertQueue = await channel.assertQueue('', { exclusive: true })
  channel.bindQueue(assertQueue.queue, exchange, '')
   channel.consume(assertQueue.queue, consume)
}

```

Although to me this looks like extra code with no benefit.

Pay attention to what you declare as `async` when currying.

Reply:
thanks, @lautarodragan, I like the first approach, it's more clear